File: day27/updateMatrix.py
'''
给定一个由 0 和 1 组成的矩阵，找出每个元素到最近的 0 的距离。
两个相邻元素间的距离为 1 。
'''

class Solution:
    def updateMatrix(self, matrix: List[List[int]]) -> List[List[int]]:
        m, n = len(matrix), len(matrix[0])
        q = collections.deque()

        for i in range(m):
            for j in range(n):
                if matrix[i][j] == 0:
                    q.append((i, j))
                else:
                    matrix[i][j] = -1

        # 广度优先搜索
        while q:
            i, j = q.popleft()
            for ni, nj in [(i - 1, j), (i + 1, j), (i, j - 1), (i, j + 1)]:
                if 0 <= ni < m and 0 <= nj < n and matrix[ni][nj] == -1:
                    matrix[ni][nj] = matrix[i][j] + 1
                    q.append((ni, nj))
        return matrix

File: day27/updateMatrix.java
class Solution {
    public int[][] updateMatrix(int[][] matrix) {
        int m = matrix.length;
        int n = matrix[0].length;

        Queue<int[]> queue = new LinkedList<>();
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (matrix[i][j] == 0) {
                    queue.offer(new int[]{i, j});
                } else {
                    matrix[i][j] = -1;
                }
            }
        }
        
        int[][] orient = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
        while (!queue.isEmpty()) {
            int[] point = queue.poll();
            int i = point[0], j = point[1];
            for (int k = 0; k < 4; k++) {
                int ni = i + orient[k][0], nj = j + orient[k][1];
                if (ni >= 0 && ni < m && nj >= 0 && nj < n && matrix[ni][nj] == -1) { 
                    matrix[ni][nj] = matrix[i][j] + 1;
                    queue.offer(new int[]{ni, nj});
                }
            }
        }

        return matrix;
    }
}
